trieste/acquisition/function/greedy_batch.py

```diff
@@ -351,3 +355,268 @@ def __call__(self, x: TensorType) -> TensorType:
+            base_acquisition_function_builder = base_acquisition_function_builder.using(OBJECTIVE)
+
+        self._builder = base_acquisition_function_builder
+        self._fantasize_method = fantasize_method
+
+    def prepare_acquisition_function(
+        self,
+        models: Mapping[str, ProbabilisticModel],
+        datasets: Optional[Mapping[str, Dataset]] = None,
+        pending_points: Optional[TensorType] = None,
+    ) -> AcquisitionFunction:
+        """
+
+        :param models: The models over each tag.
+        :param datasets: The data from the observer (optional).
+        :param pending_points: Points already chosen to be in the current batch (of shape [M,D]),
+            where M is the number of pending points and D is the search space dimension.
+        :return: An acquisition function.
+        """
+        if pending_points is not None:
+            tf.debugging.assert_rank(pending_points, 2)
+
+            fantasized_data = {
+                tag: self._generate_fantasized_data(model, pending_points)
+                for tag, model in models.items()
+            }
+            models = {
+                tag: _fantasize_model(model, fantasized_data[tag]) for tag, model in models.items()
+            }
+            if datasets is None:
+                datasets = fantasized_data
+            else:
+                datasets = {tag: data + fantasized_data[tag] for tag, data in datasets.items()}
+        return self._builder.prepare_acquisition_function(models, datasets)
+
+    def _generate_fantasized_data(
+        self, model: ProbabilisticModel, pending_points: TensorType
+    ) -> Dataset:
+        if self._fantasize_method == "KB":
+            fantasized_obs, _ = model.predict(pending_points)
+        else:
+            fantasized_obs = model.sample(pending_points, num_samples=1)  # TODO check dim
+
+        return Dataset(pending_points, fantasized_obs)
+
+
+def _fantasize_model(
+    model: ProbabilisticModel, fantasized_data: Dataset
+) -> _fantasized_model | ModelStack:
+    if isinstance(model, ModelStack):
+        observations = tf.split(fantasized_data.observations, model._event_sizes, axis=-1)
+        fmods = []
+        for mod, obs, event_size in zip(model._models, observations, model._event_sizes):
+            fmods.append(
+                (
+                    _fantasized_model(mod, Dataset(fantasized_data.query_points, obs)),
+                    event_size,
+                )
+            )
+        return ModelStack(*fmods)
+    else:
+        return _fantasized_model(model, fantasized_data)
+
+
+class _fantasized_model(ProbabilisticModel):
+    """
+    Creates a new model from an existing one and additional data.
+    This new model posterior is conditioned on both current model data and the additional one.
+    """
+
+    def __init__(self, model: ProbabilisticModel, fantasized_data: Dataset):
+        """
+        :param model: a model, must be of :class:`~FantasizedGPRModel`
```

Comment:
```suggestion
        :param model: a model, must be of :class:`~FastUpdateModel`
```

Reply:
not sure if we need to use `:class:`~trieste.models.interfacesFastUpdateModel``. here for the doc building or Sphynx picks up the import, would be good to build the docs and check before merging in